src/conversions/std/time.rs

```diff
@@ -0,0 +1,153 @@
+#![cfg(not(Py_LIMITED_API))]
+
+use crate::types::{PyDelta, PyDeltaAccess};
+use crate::{FromPyObject, IntoPy, PyAny, PyObject, PyResult, Python, ToPyObject};
+use std::convert::TryInto;
+use std::time::Duration;
+
+const DAY_SECONDS: u64 = 60 * 60 * 24;
+
+impl ToPyObject for Duration {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        let total_secs = self.as_secs();
+        // Total number of days
+        let days = total_secs / DAY_SECONDS;
+        // Remainder of seconds
+        // .try_into() converts i64 to i32, but this should never overflow
+        // since it's at most the number of seconds per day
+        let secs = (total_secs - days * DAY_SECONDS).try_into().unwrap();
+        // Fractional part of the duration
+        // Same for the conversion from i64 to i32
+        let micros = self.subsec_micros().try_into().unwrap();
+
+        // We do not need to check i64 to i32 cast from rust because
+        // python will panic with OverflowError.
+        // We pass true as the `normalize` parameter since we'd need to do several checks here to
+        // avoid that, and it shouldn't have a big performance impact.
+        let delta = PyDelta::new(py, days.try_into().unwrap_or(i32::MAX), secs, micros, true)
+            .expect("Failed to construct delta");
+        delta.into()
+    }
+}
+
+impl IntoPy<PyObject> for Duration {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        ToPyObject::to_object(&self, py)
+    }
+}
+
+impl FromPyObject<'_> for Duration {
+    fn extract(ob: &PyAny) -> PyResult<Duration> {
+        let delta: &PyDelta = ob.downcast()?;
+        // Python size are much lower than rust size so we do not need bound checks.
+        // 0 <= microseconds < 1000000
+        // 0 <= seconds < 3600*24
+        // -999999999 <= days <= 999999999
+        let days: u64 = delta.get_days().try_into().unwrap();
+        let seconds: u64 = delta.get_seconds().try_into().unwrap();
+        let microseconds: u32 = delta.get_microseconds().try_into().unwrap();
+        Ok(Duration::new(
+            days * DAY_SECONDS + seconds,
+            microseconds * 1000,
+        ))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::panic;
+
+    #[test]
+    fn test_std_duration_topyobject() {
+        // Utility function used to check different durations.
+        // The `name` parameter is used to identify the check in case of a failure.
+        let check = |name: &'static str, delta: Duration, py_days, py_seconds, py_ms| {
```

Comment:
Personally I prefer to have this as a separate helper function `_test_duration_to_py` and then each `check` call becomes its own function e.g. `test_delta_zero_to_py`. Not a huge thing, just makes it slightly easier to see each test case and what their intended purpose is.